New consistency tests introduced by CASSANDRA-9102

stef1927 opened the pull request:
There are the tests for [CASSANDRA-9102] (https://issues.apache.org/jira/browse/CASSANDRA-9102) : Consistency levels such as non-local quorum need better tests.

aboudreault commented on consistency_test.py:
```diff
@@ -1,149 +1,601 @@
+        return cl == ConsistencyLevel.LOCAL_QUORUM or \
+               cl == ConsistencyLevel.LOCAL_ONE or \
+               cl == ConsistencyLevel.LOCAL_SERIAL
+
```
return (cl == ConsistencyLevel.LOCAL_QUORUM or
            cl == ConsistencyLevel.LOCAL_ONE or
            cl == ConsistencyLevel.LOCAL_SERIAL)
nit: I'd would prefer this way rather than \

aboudreault commented on consistency_test.py:
```diff
@@ -1,149 +1,601 @@
+    def _is_conditional(self, cl):
+        return cl == ConsistencyLevel.SERIAL or \
+               cl == ConsistencyLevel.LOCAL_SERIAL
+
```
same comment than above

ptnapoleon commented on consistency_test.py:
```diff
@@ -1,149 +1,601 @@
+            for i in xrange(0, len(rf_factors)):
+                if num_nodes_alive[i] < self._required_nodes(cl, rf_factors, i):
+                    return False
+                return True
```
This line is intended one too far, right? We'd be exiting after the first iteration of the for loop

ptnapoleon commented on consistency_test.py:
```diff
@@ -1,149 +1,601 @@
+    """
+    def _test_simple_strategy(self, combinations):
+        """
+        Helper test function for a sinle data center: invoke _test_insert_query_from_node() for each node
```
s/sinle/single

mambocab commented on consistency_test.py:
```diff
@@ -1,149 +1,601 @@
+                self.cluster.nodelist()[node].stop()
+                nodes_alive[i] = nodes_alive[i] - 1
+
+    def _test_insert_query_from_node(self, session, dc_idx, rf_factors, num_nodes_alive, write_cl, read_cl, serial_cl=None, check_ret=True):
```
This method should provide a little more help in tracking down a failure if one happens. Since this method is called in a loop, e.g. in `test_simple_strategy`, a failure will just look like "FAIL: `test_simple_strategy`", which doesn't specify the particular consistency levels the test failed at. I think a `debug` call that prints the relevant parameters would be enough.

aboudreault replied:
+1. I think adding some more debug to exactly what exactly is tested would be simple and good enough there.

stef1927 replied:
We have already this debug line at the beginning of the method:
```python
self.log("Connected to %s for %s/%s/%s" %
            (session.cluster.contact_points, self._name(write_cl), self._name(read_cl), self._name(serial_cl)))
```
We also have debug calls in the loops calling this method. Where exactly would you like to see more information?

mambocab replied:
You're right, I totally misread that line. This debug line should be plenty, sorry about that!

stef1927 replied:
No problem, thanks for confirming! :)

ptnapoleon commented on consistency_test.py:
```diff
@@ -1,149 +1,601 @@
-        for n in xrange(0, 100):
-            insert_c1c2(session, n, ConsistencyLevel.QUORUM)
-        assert_unavailable(query_c1c2, session2, 100, ConsistencyLevel.THREE)
+    def _restart_cluster(self):
```
I don't see where else this method is called. Is it misnamed? It isn't really restarting a cluster, only booting a previously stopped one.

ptnapoleon commented on consistency_test.py:
```diff
@@ -1,149 +1,601 @@
+            self.sessions = []
+            self.sessions.append(session)
+            for node in cluster.nodelist()[1:]:
+                self.sessions.append(self.patient_cql_connection(node, self.ksname))
```
Should these be patient, exclusive connections? Otherwise, they're all essentially equivalent, so I don't see why we would use multiple.

ptnapoleon commented on consistency_test.py:
```diff
@@ -1,149 +1,601 @@
+            ConsistencyLevel.LOCAL_ONE : 1,
+        }[cl]
+
+    def _should_suceed(self, cl, rf_factors, num_nodes_alive, current):
```
s/suceed/succeed

aboudreault commented on consistency_test.py:
```diff
@@ -1,149 +1,601 @@
+
+    def insert_user(self, session, userid, age, consistency, serial_consistency=None):
+        text = "INSERT INTO users (userid, firstname, lastname, age) VALUES (%d, 'first%d', 'last%d', %d) %s" \
+            % (userid, userid, userid, age, "IF NOT EXISTS" if serial_consistency else "")
```
***NO IMMEDIATE MODIFICATION REQUIRED***

This is comment for future PRs. Since Python 3 will eventually become default in future OS distributions, we try to use as many as possible Py3 habits in our new code, ie. the string formatting. The formatting use the "".format() function.

https://docs.python.org/3.5/library/string.html#format-string-syntax

***NO IMMEDIATE MODIFICATION REQUIRED***

ptnapoleon replied:
What Alan is saying, is that formatting with `%` here is fine, but in future PRs, if you could use the python3 way, that would be preferable, please.

stef1927 replied:
Got it, thanks!

mambocab commented on consistency_test.py:
```diff
@@ -1,149 +1,601 @@
+    def read_counter(self, session, id, consistency):
+        statement = SimpleStatement("SELECT c from counters WHERE id = %d" % (id,), consistency_level=consistency)
+        res = rows_to_list(session.execute(statement))
+        return res[0][0] if len(res) > 0 else 0
```
Nit: It's more idiomatic to write `if res` than `if len(res) > 0`.

I'm not clear on why returning 0 is ok if no result is returned. Maybe add a comment to explain.

mambocab commented:
Great tests. I agree with Alan and Philip's comments; when the comments are addressed I'll have a look at the changes.

stef1927 commented:
Thanks for your comments. I pushed a new commit and I believe I addressed all comments requiring modification, except one for which I asked a question inline above. Please let me know if I missed something.

ptnapoleon commented:
I'm +1. @mambocab or @aboudreault ?

aboudreault commented:
Yep +1. Thanks @stef1927  for the great work!

mambocab commented:
+1, thanks!
